<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

* Cement
* Blast furnace slag
* Fly ash
* Water
* Superplasticizer
* Coarse aggregate
* Fine aggregate

#### Find strength of concrete

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras

In [3]:
df = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
df.head()


,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age,Strength
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28,79.99
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28,61.89
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270,40.27
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365,41.05
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360,44.30


#### So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

In [5]:
df.shape

(1030, 9)

In [6]:
df.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [16]:
#SEPRATE DATA INTO FEATURES AND LABELS

concrete_Column = df.columns

input = df[concrete_Column[concrete_Column != 'Strength']]
output = df['Strength']

In [17]:
input.head()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360



<img src="https://spreadsheetweb.com/wp-content/uploads/2020/07/How-to-normalize-data-in-Excel-00.png" width="500">

In [21]:
input_norm =(input - input.mean())//input.std()
input_norm.head()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age
0,2.0,-1.0,-1.0,-1.0,-1.0,0.0,-2.0,-1.0
1,2.0,-1.0,-1.0,-1.0,-1.0,1.0,-2.0,-1.0
2,0.0,0.0,-1.0,2.0,-2.0,-1.0,-3.0,3.0
3,0.0,0.0,-1.0,2.0,-2.0,-1.0,-3.0,5.0
4,-1.0,0.0,-1.0,0.0,-2.0,0.0,0.0,4.0


why we normalize the data

We normalize data in neural networks for several critical reasons:

* Equal Feature Importance

Your concrete features have very different scales (e.g., cement might be 100-500 kg/m³, while superplasticizer is 0-30 kg/m³)
Without normalization, features with larger values dominate the learning process

* Faster Convergence

Neural networks train much faster when all inputs are on a similar scale (typically mean=0, std=1)
Gradient descent converges more quickly and smoothly

* Prevents Gradient Issues

Large input values can cause exploding gradients
Small input values can cause vanishing gradients
Normalized data keeps gradients in a healthy range
* Better Weight Initialization

Neural network weights are typically initialized for normalized inputs
Unnormalized data can make initial weights ineffective

* Numerical Stability

Prevents overflow/underflow errors during computation
Makes the optimization landscape smoother and easier to navigate

In [22]:
n_cols = input_norm.shape[1]

In [23]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input

### Build a Neural Network
